draft-jennings-moq-secure-objects.md

```diff
@@ -201,326 +190,248 @@ name but different data.
+> problematic for other reasons (namely object name uniqueness).  In any case,
+> these variants can be distinguished by KID, and the KID will have to be
+> carried in the protected object payload.
+
+> **NOTE:** You could also use the track names to derive keys off of a master
+> key, something like `track_base_key = HKDF(track_id, master_key)`.  This could
+> be useful, e.g., if the `master_key` is exported from MLS.  SFrame will also
+> derive different base keys for different KID values, so in principle, you
+> could use the KID to distinguish tracks.  But as discussed above, we probably
+> want to use the KID for other purposes.
+
+## Mapping Between SFrame and MOQT
+
+MOQT secure objects are protected and unprotected by means of the analogous
+SFrame operations.  SFrame the protect operation requires KID, CTR, and metadata
+inputs, and produces an SFrame ciphertext structure as output; the unprotect
+operation takes as input an SFrame ciphertext (including KID and CTR in a
+header) and metadata.
+
+~~~ aasvg
++-+----+-+----+--------------------+--------------------+
+|K|KLEN|C|CLEN|     Key ID = 0     | Counter = Grp+Obj  | <---- Synthesized
++-+----+-+----+--------------------+--------------------+ -.
+|                                                       |   |
+|                                                       |   |
+|                                                       |   |
+|                                                       |   |
+|                   Encrypted Data                      |   |
+|                                                       |   +-- MOQT Object
+|                                                       |   |   Payload
+|                                                       |   |
+|                                                       |   |
++-------------------------------------------------------+   |
+|                 Authentication Tag                    |   |
++-------------------------------------------------------+ -'
+~~~
+{: #fig-sframe-ciphertext title="An SFrame, as it relates to MOQT" }
+
+For protect/unprotect of MOQT objects, the KID, CTR and metadata fields are
+computed from the MOQT object:
+
+* `KID = 0`
+* `CTR = (group_id << 32) | (object_id & 0xFFFFFFFF)`
+* `metadata = subscribe_id || ... || object_status` (i.e., the entirety of the
+  message aside from the payload)
+
+In order for this mapping to be one-to-one (which is required to avoid nonce
+reuse), both the Group ID and Object ID fields in the MOQT object MUST be less
+than 2<sup>32</sup>.
+
+> **NOTE:** If this limitation is uncomfortable, it can be mitigated in a few
+> ways, for example:
+> * having a key management scheme that allows for a track to be re-keyed
+> * Allowing the Group ID and Object ID subfields of the CTR value to have
+>   different sizes, so that for example an audio stream could have 0 object ID
+>   bits (thus requiring the object ID to always be zero) and 64 bits of group
+>   ID.
```

Comment:
Done in a nieve way, this could create some security problems, this this would need to be pretty clearly defined. I agree limiting group ID to 32 bits will probably not fly but I don't think there would be issue with limiting Object ID to 30 or 32 bits.

Reply:
on call you had idea of top few bits indicate the cut point, then can have variable side object and group as long as total is less that about 62 bits